### PR TITLE
Create a FeatureEnableService so Layout Preview tab appears in IntelliJ

### PR DESCRIPTION
**Pull request by jbarr21**
This diffs adds a `FeatureEnableService` to the IdeaBuck plugin so that IntelliJ displays the Layout Preview "Design" tab for Android layout XML files.

Without "Design" tab below XML editor:
layout-xml-without-tabs

With "Design" tab below XML editor that provides access to Android Layout Preview:
layout-xml-with-tabs

### CODE REVIEW

**Comment by jbarr21** [comment]
> I'm trying to test this change, but I can't see which feature it enables. Can you provide some screenshots, please?

this re-enables the extra "Design" tab when editing Android Layout XML files. as of IntelliJ 2017.2, XML layouts won't display the tab without this change. screenshots added to the description

File: third-party/java/intellij/BUCK [hunk 1]
```diff
@@ -48,3 +49,8 @@ prebuilt_jar(
     name = "jdom",
     binary_jar = "jdom.jar",
 )
+
+prebuilt_jar(
+    name = "android",
+    binary_jar = "android.jar",
```
Comment (jbarr21) [inline third-party/java/intellij/BUCK]: I got this JAR from IntelliJ installation. it was needed so that we could find the `FeatureEnableService` symbol

Comment (jbarr21) [inline third-party/java/intellij/BUCK]: yes, it's from an IntelliJ installation. it does not need to be a specific version.

@styurin which piece of that do you want in a separate PR?

Comment (styurin) [inline third-party/java/intellij/BUCK]: > which piece of that do you want in a separate PR?

Updating `get_jars.sh` and adding a new jar.

**Comment by styurin** [comment]
> this re-enables the extra "Design" tab when editing Android Layout XML files. as of IntelliJ 2017.2, XML layouts won't display the tab without this change. screenshots added to the description

Do you use aggregation of modules with Android resources (option `intellij.aggregate_android_resource_modules` in `.buckconfig`)? This is the only reason I can think of why IntelliJ isn't showing that tab.

If you don't use this option, can you provide a sample project where this tab isn't shown?

**Comment by jbarr21** [comment]
>> this re-enables the extra "Design" tab when editing Android Layout XML files. as of IntelliJ 2017.2, XML layouts won't display the tab without this change. screenshots added to the description

> Do you use aggregation of modules with Android resources (option intellij.aggregate_android_resource_modules in .buckconfig)? This is the only reason I can think of why IntelliJ isn't showing that tab.
>
> If you don't use this option, can you provide a sample project where this tab isn't shown?

The .buckconfig option has no effect on the issue.

Any project (I used the [OkBuck sample app](https://github.com/uber/okbuck)), when opened with the IntelliJ 2017.2 build, does not display the editor tab. This is because the [condition for the Layout Preview Editor](https://github.com/JetBrains/android/blob/master/designer/src/com/android/tools/idea/uibuilder/editor/NlEditorProvider.java#L46) to be displayed fails since the [GradleFeatureEnableService](https://github.com/JetBrains/android/blob/master/android/src/com/android/tools/idea/gradle/project/GradleFeatureEnableService.java#L25) requires the project to be [built with Gradle](https://github.com/JetBrains/android/blob/master/android/src/com/android/tools/idea/gradle/project/GradleProjectInfo.java#L68).

This PR adds a FeatureEnableService extension that checks if the project was built with Buck, and enables the Layout Preview editor tab.

File: third-party/java/intellij/BUCK (hunk 1 above)
Comment (jbarr21) [inline third-party/java/intellij/BUCK]: ok, will do, but then this PR won't compile (since no android.jar), is that OK?

Comment (jbarr21) [inline third-party/java/intellij/BUCK]: Added in #1448

**Comment by styurin** [comment]
This is failing to build because `FeatureEnableService` was added in 2017.2 and I updated libraries from 2017.1. This merge this PR we would have to bump ideabuck's min version to 2017.2, but right now we are using 2017.1 internally. I think the only way to proceed here is to wait for us to update to 2017.2 and then we can merge this fix.

cc @mkillianey

**Comment by mkillianey** [comment]
I'm working on a change to our chef scripts so that we can install the old
version in 2017.1 and the new version in 2017.2 on the same machine.  I
hope to have that change ready by next week.

**Comment by styurin** [comment]
@jbarr21 could you update this PR with a new android.jar from IntelliJ 2017.2? I want to review this code while Mick is working on his fix.

**Comment by styurin** [comment]
Can you also update IntelliJ configuration to allow to build the plugin from IntelliJ?

**Comment by styurin** [comment]
OMG, it works with merging android resources! Thanks a lot :)

**Comment by styurin** [comment]
By the way, android.jar should be added as a provided dependency in ideabuck.iml to avoid pulling stripped jar.

**Comment by jbarr21** [comment]
@styurin can you please clarify the android.jar via ideabuck.iml comment? shouldn't the dependency be provided via the buck configuration, i'll just update the script to not stub the android.jar?

**Comment by jbarr21** [comment]
@styurin just wanted to check on the status of your IntelliJ installs. is this PR still blocked on that?

as for building the plugin from IntelliJ, that sounds outside of the scope of this PR

**Comment by styurin** [comment]
@mkillianey can you please comment if we are ready to bump the minimum version for buck plugin to 2017.2?

> can you please clarify the android.jar via ideabuck.iml comment?

When building and testing plugin inside IntelliJ (using IntelliJ build-in functionality, not buck build), the plugin jar contains stripped version of android.jar and, since the content of the stripped jar is not working, the plugin is not working. If dependency type on android.jar is "provided" then android.jar is not included in the plugin jar and the plugin is using classes provided by InteliiJ.

**Comment by jbarr21** [comment]
IntelliJ 2017.2.4 picks up this fix for the issue, though I think it's still a good idea to continue this PR in case their legacy logic changes in the future https://github.com/JetBrains/android/commit/09d47f3a608659723b98f46505b2c35d32436048

File: src/com/facebook/buck/intellij/ideabuck/META-INF/plugin.xml [hunk 2]
```diff
@@ -207,6 +209,10 @@
     <runLineMarkerContributor language="JAVA" implementationClass="com.facebook.buck.intellij.ideabuck.actions.select.SelectedTestRunLineMarkerContributor"/>
   </extensions>
 
+  <extensions defaultExtensionNs="com.android.project">
+    <featureEnableService implementation="com.facebook.buck.intellij.ideabuck.proj.BuckFeatureEnableService"/>
```
Comment (mkillianey) [inline src/com/facebook/buck/intellij/ideabuck/META-INF/plugin.xml]: Q:  Does using this mean that you need to update the `since-build` on line 144?

**Comment by styurin** [comment]
I think this change is required for a project generated by `buck project` when Android resource module aggregation is enabled because such projects are not legacy.

What this PR needs at this point:

- bump minimal version to 2017.2 (a version where `FeatureEnableService` first appeared
- upload a version of `android.jar` from 2017.2. The version in buck repo is from 2017.1
